lib/src/rules/prefer_enums.dart

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+
+import '../analyzer.dart';
+import '../ast.dart';
+import '../util/dart_type_utilities.dart';
+
+const _desc = r'Prefer enums over classes that behave like enums.';
+
+const _details = r'''
+Classes that look like enumerations should be declared as `enum`s.
+
+**DO** use enums where appropriate.
+
+Candidates for enums are classes that:
+  * are concrete,
+  * are private or have only private generative constructors,
+  * have two or more static const fields with the same type as the class,
+  * have generative constructors that are only invoked at the top-level of the
+    initialization expression of these static fields,
+  * do not define `hashCode`, `==` or `index`,
```

Comment:
So... technically `index` might be OK if it's being used in a migratable way.  The open-question is whether this is something the lint should be aware of.